feat: 🎸 add validation handler (middleware) (#39524)

permil-successive opened the pull request:
add validation handler (middleware) which validate input variables in
the request as per config file.

✅ Closes: #39524

@nxvinay  @Prabal555

permil-successive committed:
feat: 🎸 add validation handler (middleware)

add validation handler (middleware) which validate input variables in
the request as per config file.

✅ Closes: #39524

permil-successive committed:
fix: 🐛 fix throw error on unwanted variable in required

Prabal555 commented on src/controllers/trainee/routes.ts:
```diff
@@ -1,13 +1,17 @@
+.put(validationHandler(validationConfig.update), update)
+.delete(validationHandler(validationConfig.delete), deleteFunction);
+
+traineeRoutes.delete('/:id', validationHandler(validationConfig.delete), deleteFunction);
```
It would be better to destructor the configuration at the top. like you have for the controller.

Prabal555 commented on src/libs/errorHandler.ts:
```diff
@@ -4,17 +4,36 @@ import { Request, Response, NextFunction } from 'express';
+      message: err.message,
+      status: err.code,
+      timestamp: new Date().toUTCString()
+    };
```
We can have a separate method for formating the error messages

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,120 @@
+  return regex.test(stringToValidate);
+}
+
+export default (config) => (req: Request, res: Response, next: NextFunction) => {
```
the function is very long should be broken down into smaller function especially into functions that validate for each keys

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,120 @@
+    validationConfig.in = validationConfig.in || [ 'body' ];
+    let isError = false;
+
+    validationConfig.in.forEach((element) => {
```
if some inserts an incorrect value in the 'in' key of the configuration then this will break the code

permil-successive committed:
refactor: 💡 beautify the code

permil-successive committed:
refactor: 💡 add comments and break longer functions

add comments, break longer functions, merge import/export in single
statement wherever required

permil-successive committed:
fix: 🐛 remove extra supplied variable

permil-successive committed:
fix: 🐛 fix regex bug

Prabal555 commented on src/controllers/trainee/routes.ts:
```diff
@@ -1,13 +1,18 @@
+.put(validationHandler(updateConfig), update)
+.delete(validationHandler(deleteConfig), deleteFunction);
+
+traineeRoutes.delete('/:id', validationHandler(validationConfig.delete), deleteFunction);
```
two delete methods and in the second delete method destructed method is not used

Prabal555 commented on src/controllers/trainee/validations.ts:
```diff
@@ -0,0 +1,77 @@
+      custom: (dataToUpdate: any, errorMessage: IError): boolean => {
+        if (dataToUpdate.id === undefined) {
+          errorMessage.message = 'id is required in dataToUpdate';
+          return false;
```
redundant id values

Prabal555 commented on src/controllers/trainee/validations.ts:
```diff
@@ -0,0 +1,77 @@
+      default: 0,
+      number: true,
+      in: ['query'],
+      errorMessage: 'Skip is invalid',
```
cases of error message 'Skip' is not consistent with the key they denote 'skip', it should be the same every where

Prabal555 commented on src/controllers/trainee/validations.ts:
```diff
@@ -0,0 +1,77 @@
+          errorMessage.message = 'id is not start with A';
+          errorMessage.code = '401';
+          return false;
+        } else {
```
Please avoid calling changing the values for arguments in place please use pure functions instead

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+  if (errorMessages.length > 0)
+    next(errorMessages);
+  else
+    next();
```
no need to use else here

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+function validateValidationConfig (validationConfig): void {
+
+  // assigning default config
+  validationConfig.required = validationConfig.required === undefined ? true : validationConfig.required;
+  validationConfig.in = validationConfig.in || [ 'body' ];
```
donnot assign values to the arguments user pure functions instead

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    (toValidate === undefined || toValidate === null ||
+      (typeof toValidate === 'string' && toValidate.trim() === '')
+    )
+  ) {
```
code can be made easier to read and understand

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    return true;
+  } else {
+    return false;
+  }
```
no need for else

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    return true;
+  } else {
+    return false;
+  }
```
no need for else

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    return true;
+  } else {
+    return false;
+  }
```
no need for else

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    return true;
+  } else {
+    return false;
+  }
```
no need for else

Prabal555 commented on src/libs/validationHandler.ts:
```diff
@@ -0,0 +1,241 @@
+    const validationConfig = config[key];
+    let isError = false;
+
+    validateValidationConfig(validationConfig);
```
should be encapsulated in try catch if expected to throw a value

Prabal555 commented:
Getting 401 unauthorized http status, **please use the correct response code**
![image](https://user-images.githubusercontent.com/45360184/73875378-383e3680-487b-11ea-9fc9-cd5cf35cfc39.png)

**What is the purpose of this code ?**
`
this.app.get('/body-parser', (req: express.Request, res: express.Response): void => {
      console.log('user send data : ');
      console.log(req.body);
      res.send('Ok');
});
` 

**No index file in the controller folder**

**The format for the timestamps being returned from the application should be the same**
![image](https://user-images.githubusercontent.com/45360184/73876673-6755a780-487d-11ea-8975-672a001c9c2c.png)

**A single helper method should have been used to create a response format **
`    const responseMessage: IResponse = {
      message: 'Record inserted successfully.',
      status: 'OK',
      timestamp: new Date().toISOString(),
      data: traineeData
    };`

**Validation of the type is not correct for dataToUpdate**
![image](https://user-images.githubusercontent.com/45360184/73877455-b18b5880-487e-11ea-9687-099a6f228711.png
![image](https://user-images.githubusercontent.com/45360184/73877503-c7008280-487e-11ea-910d-e5521306636f.png)

**Validations not being handled properly**
![image](https://user-images.githubusercontent.com/45360184/73877700-1fd01b00-487f-11ea-8317-b8d9de5f992e.png)

**Redundant ID field in update**

**Incorrect response**
![image](https://user-images.githubusercontent.com/45360184/73877984-b7ce0480-487f-11ea-8481-eb6d5f3ef448.png)

permil-successive committed:
refactor: 💡 do changes as request in PR#10

remove unwanted route from server.ts, add a single response helper, fix
trainee delete route, fix custom validation function of trainee config,
fix response status code, use try/catch and pure function in
validationHandler.ts

✅ Closes: #39524

permil-successive committed:
fix: 🐛 fix bug in validation handler

for string datatype
